### PR TITLE
feat(cni): generate BaseSvc cache and push service cluster ip flow

### PR DESCRIPTION
**Pull request by qingwusunny**

### CODE REVIEW

File: pkg/agent/controller/proxy/controller.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+		if err := r.updateService(ctx, &svc, old); err != nil {
+			klog.Infof("Failed to update service: %v, err: %s", req.NamespacedName, err)
+			return ctrl.Result{}, err
+		}
+		klog.Infof("Success update service: %v", req.NamespacedName)
+	} else {
+		if err := r.createService(ctx, &svc); err != nil {
+			klog.Infof("Failed to create service: %v, err: %s", req.NamespacedName, err)
+		}
+		klog.Infof("Success create service: %v", req.NamespacedName)
```
Comment (KeyongSun) [inline pkg/agent/controller/proxy/controller.go:57]: create failed and will also print success message?
when failed need to return err

File: pkg/agent/controller/proxy/controller.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	newSvcCache := svcToBaseSvc(newSvc)
+	if newSvcCache == nil {
+		return fmt.Errorf("the transferred BaseSvc from service %v is nil", *newSvc)
+	}
+
+	// todo ProcessIPChange
+	// todo ProcessPortChange
+	// todo ProcessSessionAffinityChange
+
+	r.baseSvcCache.AddOrUpdate(newSvcCache)
```
Comment (KeyongSun) [inline pkg/agent/controller/proxy/controller.go:108]: check return result?

File: pkg/agent/controller/proxy/controller.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	return nil
+}
+
+func (r *Reconcile) deleteService(ctx context.Context, svcCache *BaseSvc) error {
+	if svcCache == nil {
+		return nil
+	}
+	// todo delete flow
+
+	r.baseSvcCache.Delete(svcCache)
```
Comment (KeyongSun) [inline pkg/agent/controller/proxy/controller.go:119]: Ditto

Comment (qingwusunny) [reply pkg/agent/controller/proxy/controller.go]: the func has no return error

File: pkg/agent/controller/proxy/controller.go (hunk 2 above)
Comment (qingwusunny) [reply pkg/agent/controller/proxy/controller.go]: this func has no return

File: Makefile [hunk 4]
```diff
@@ -45,7 +45,8 @@ agent-uuid:
 	cat /proc/sys/kernel/random/uuid > /var/lib/everoute/agent/name
 
 test: agent-uuid
-	go test ./plugin/... ./pkg/...
+	go test ./pkg/agent/controller/proxy/...
+# go test ./plugin/... ./pkg/...
```
Comment (wu-cl) [inline Makefile:49]: add plugin test back to unitest

Comment (qingwusunny) [reply Makefile]: done

File: cmd/everoute-controller/main.go [hunk 5]
```diff
@@ -149,8 +149,9 @@ func main() {
 
 		if enableProxy {
 			if err = (&k8s.EndpointsReconcile{
-				Client: mgr.GetClient(),
-				Scheme: mgr.GetScheme(),
+				APIReader: mgr.GetAPIReader(),
```
Comment (echkenluo) [inline cmd/everoute-controller/main.go:152]: Add api reader to read directly from apiserver? why not read it from cache, it has some special case ？

Comment (qingwusunny) [reply cmd/everoute-controller/main.go:152]: when list servicePort resource, the cache doesn't sync in time

File: pkg/agent/controller/proxy/controller.go [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+	}
+	return nil
+}
+
+func (r *Reconcile) addClusterIP(ip string, baseSvc *proxycache.BaseSvc) error {
+	if baseSvc == nil {
+		return fmt.Errorf("missing service base info for add cluster IP: %s", ip)
+	}
+	dpNatBrs := r.DpMgr.GetNatBridges()
+	for i := range dpNatBrs {
```
Comment (echkenluo) [inline pkg/agent/controller/proxy/controller.go:242]: Are there multiple nat bridges?

File: pkg/agent/datapath/cache/svc_ovs_info.go [hunk 7]
```diff
@@ -57,23 +57,45 @@ func (s *SvcOvsInfo) DeleteGroup(portName string) {
 	log.Debugf("Delete the group id of port name %s", portName)
 }
 
-func (s *SvcOvsInfo) GetLBFlowIDsByIP(lbIP string) []*ofctrl.Flow {
+func (s *SvcOvsInfo) GetLBFlow(lbIP, portName string) *ofctrl.Flow {
```
Comment (echkenluo) [inline pkg/agent/datapath/cache/svc_ovs_info.go:60]: Is there such a case：different endpoints of the same service use different port？ Is k8s support this config？

File: pkg/agent/datapath/cache/svc_ovs_info.go [hunk 8]
```diff
@@ -57,23 +57,45 @@ func (s *SvcOvsInfo) DeleteGroup(portName string) {
 	log.Debugf("Delete the group id of port name %s", portName)
 }
```
Comment (echkenluo) [inline pkg/agent/datapath/cache/svc_ovs_info.go:58]: L22 ~ L25 comment flowID ---> flow

File: pkg/agent/datapath/natBridge.go [hunk 9]
```diff
@@ -153,35 +155,151 @@ func (n *NatBridge) RemoveLocalEndpoint(endpoint *Endpoint) error {
...
+
+	if protocol == corev1.ProtocolUDP {
+		newFlow, err := n.serviceLBTable.NewFlow(ofctrl.FlowMatch{
+			Priority:       MID_MATCH_FLOW_PRIORITY,
+			Ethertype:      PROTOCOL_IP,
+			IpProto:        PROTOCOL_UDP,
+			IpDa:           ipDa,
+			IpDaMask:       &OneIPMask,
+			UdpDstPort:     uint16(port),
+			UdpDstPortMask: OnePortMask,
```
Comment (echkenluo) [inline pkg/agent/datapath/natBridge.go:268]: OnePortMask ---> matchFullBitMask ?

**Review by echkenluo — APPROVED** [review]

File: pkg/agent/datapath/cache/svc_ovs_info.go (hunk 7 above)
Comment (qingwusunny) [reply pkg/agent/datapath/cache/svc_ovs_info.go:60]: when a service port's targetPort is a named port, different pod may be correspond  different port

File: pkg/agent/datapath/cache/svc_ovs_info.go (hunk 8 above)
Comment (qingwusunny) [reply pkg/agent/datapath/cache/svc_ovs_info.go:58]: done

File: pkg/agent/datapath/natBridge.go (hunk 9 above)
Comment (qingwusunny) [reply pkg/agent/datapath/natBridge.go:268]: done

**Review by echkenluo — APPROVED** [review]

**Review by KeyongSun — APPROVED** [review]

**Pull request merged**